src/imageruler/imageruler.py

```diff
@@ -0,0 +1,620 @@
+
+    Args:
+        wrapped_x: The wrapped bool-typed rank-2 array containing the features.
+        length_scale: The length scale for which violations are sought.
+        periodic: Specifies which of the two axes are to be regarded as periodic.
+        ignore_scheme: Specifies what pixels are ignored when detecting violations.
+
+    Returns:
+        The array containing violations.
+    """
+    x = wrapped_x.array
+    kernel = kernel_for_length_scale(length_scale)
+    violations_solid = x & ~binary_opening(x, kernel, periodic, PaddingMode.SOLID)
+
+    ignored = ignored_pixels(x, periodic, ignore_scheme)
+    violations_solid = violations_solid & ~ignored
+    return violations_solid
+
+
+def kernel_for_length_scale(length_scale: int) -> NDArray:
+    """Returns an approximately circular kernel for the given `length_scale`.
+
+    The kernel has shape `(length_scale, length_scale)`, and is `True` for pixels
+    whose centers lie within the circle of radius `length_scale / 2` centered on
+    the kernel. This yields a pixelated circle, which for length scales less than
+    `3` will actually be square.
+
+    Args:
+        length_scale: The length scale for which the kernel is sought.
+
+    Returns:
+        The approximately circular kernel.
+    """
+    assert length_scale > 0
+    centers = onp.arange(-length_scale / 2 + 0.5, length_scale / 2)
+    squared_distance = centers[:, onp.newaxis] ** 2 + centers[onp.newaxis, :] ** 2
+    kernel = squared_distance < (length_scale / 2) ** 2
+    # Ensure that kernels larger than `2` can be realized with a width-3 brush.
+    if length_scale > 2:
+        kernel = binary_opening(
+            kernel,
+            kernel=PLUS_3_KERNEL,
+            periodic=(False, False),
+            padding_mode=PaddingMode.VOID,
+        )
+    return kernel
+
+
+def ignored_pixels(
+    x: NDArray,
+    periodic: Tuple[bool, bool],
+    ignore_scheme: IgnoreScheme,
+) -> NDArray:
+    """Returns an array indicating locations at which violations are to be ignored.
+
+    Args:
+        x: The array for which ignored locations are to be identified.
+        periodic: Specifies which of the two axes are to be regarded as periodic.
+        ignore_scheme: Specifies the manner in which ignored locations are identified.
+
+    Returns:
+        The array indicating locations to be ignored.
+    """
+    assert x.dtype == bool
+    if ignore_scheme == IgnoreScheme.NONE:
+        return onp.zeros_like(x)
+    elif ignore_scheme == IgnoreScheme.EDGES:
+        return x & ~binary_erosion(x, PLUS_3_KERNEL, periodic, PaddingMode.SOLID)
+    elif ignore_scheme == IgnoreScheme.LARGE_FEATURE_EDGES:
+        return x & ~erode_large_features(x, periodic)
+    else:
+        raise ValueError(f"Unknown `ignore_scheme`, got {ignore_scheme}.")
+
+
+# ------------------------------------------------------------------------------
+# Array-manipulating functions backed by `cv2`.
```

Comment:
Since we are doing a giant reorg, can we move some of the functions below this line into their own modules (e.g. morphology.py and/or array_utils.py)?